### ISSUE TITLE
error: redefinition of 'struct sha256_state'

### ISSUE THREAD

**Issue by jacknaguib** [Maiqel/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55#1]
Hi,

I am getting the same error message with both scripts (yours and TP-link's). I ran the script on the following system

kali@kali:~/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master$ **dpkg-query -s linux-headers-$(uname -r)**
Package: linux-headers-5.8.0-kali2-amd64
Status: install ok installed
Priority: optional
Section: kernel
Installed-Size: 5701
Maintainer: Kali Developers <devel@kali.org>
Architecture: amd64
Source: linux
Version: 5.8.10-1kali1
Depends: linux-headers-5.8.0-kali2-common (= 5.8.10-1kali1), linux-kbuild-5.8 (>= 5.8.10-1kali1), linux-compiler-gcc-10-x86
Description: Header files for Linux 5.8.0-kali2-amd64
 This package provides the architecture-specific kernel header files for
 Linux kernel 5.8.0-kali2-amd64, generally used for building out-of-tree
 kernel modules.  These files are going to be installed into
 /usr/src/linux-headers-5.8.0-kali2-amd64, and can be used for building
 modules that load into the kernel provided by the
 linux-image-5.8.0-kali2-amd64 package.
Homepage: https://www.kernel.org

kali@kali:~/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master$ **make**

make ARCH=x86_64 CROSS_COMPILE= -C /lib/modules/5.8.0-kali2-amd64/build M=/home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master  modules
make[1]: Entering directory '/usr/src/linux-headers-5.8.0-kali2-amd64'
  CC [M]  /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.o
In file included from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/drv_types.h:30,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.c:17:
/home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/wifi.h:1031: **warning: "IEEE80211_MAX_AMPDU_BUF" redefined**
 1031 | #define IEEE80211_MAX_AMPDU_BUF 0x40
      | 
In file included from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/osdep_service_linux.h:83,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/osdep_service.h:50,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/drv_types.h:27,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.c:17:
/usr/src/linux-headers-5.8.0-kali2-common/include/linux/ieee80211.h:1600: note: this is the location of the previous definition
 1600 | #define IEEE80211_MAX_AMPDU_BUF  0x100
      | 
In file included from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/drv_types.h:78,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.c:17:
/home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/rtw_security.h:252:8: **error: redefinition of ‘struct sha256_state’**
  252 | struct sha256_state {
      |        ^~~~~~~~~~~~
In file included from /usr/src/linux-headers-5.8.0-kali2-common/include/linux/filter.h:23,
                 from /usr/src/linux-headers-5.8.0-kali2-common/include/net/sock.h:59,
                 from /usr/src/linux-headers-5.8.0-kali2-common/include/linux/tcp.h:19,
                 from /usr/src/linux-headers-5.8.0-kali2-common/include/linux/ipv6.h:88,
                 from /usr/src/linux-headers-5.8.0-kali2-common/include/net/addrconf.h:50,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/osdep_service_linux.h:48,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/osdep_service.h:50,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/include/drv_types.h:27,
                 from /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.c:17:
/usr/src/linux-headers-5.8.0-kali2-common/include/crypto/sha.h:84:8: note: originally defined here
   84 | struct sha256_state {
      |        ^~~~~~~~~~~~
make[3]: *** [/usr/src/linux-headers-5.8.0-kali2-common/scripts/Makefile.build:286: /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master/core/rtw_cmd.o] Error 1
make[2]: *** [/usr/src/linux-headers-5.8.0-kali2-common/Makefile:1764: /home/kali/source-code/TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55-master] Error 2
make[1]: *** [/usr/src/linux-headers-5.8.0-kali2-common/Makefile:185: __sub-make] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.8.0-kali2-amd64'
make: *** [Makefile:1999: modules] Error 2

**Comment by jandd** [comment]
same here on Debian testing/bullseye with Kernel 5.9.0-4. Unfortunately the definitions in the kernel headers and in the module are different/incompatible:

in kernel header file /usr/src/linux-headers-5.9.0-4-common/include/crypto/sha.h:
```
static inline void sha256_init(struct sha256_state *sctx)
{
  sctx->state[0] = SHA256_H0;
  sctx->state[1] = SHA256_H1;
  sctx->state[2] = SHA256_H2;
  sctx->state[3] = SHA256_H3;
  sctx->state[4] = SHA256_H4;
  sctx->state[5] = SHA256_H5;
  sctx->state[6] = SHA256_H6;
  sctx->state[7] = SHA256_H7;
  sctx->count = 0;
}
```

in the driver code TP-Link-Archer-T4U-AC1300-Linux-driver-kernel-55/core/rtw_security.c:

```
static void sha256_init(struct sha256_state *md)
{
  md->curlen = 0;
  md->length = 0;
  md->state[0] = 0x6A09E667UL;
  md->state[1] = 0xBB67AE85UL;
  md->state[2] = 0x3C6EF372UL;
  md->state[3] = 0xA54FF53AUL;
  md->state[4] = 0x510E527FUL;
  md->state[5] = 0x9B05688CUL;
  md->state[6] = 0x1F83D9ABUL;
  md->state[7] = 0x5BE0CD19UL;
}
```

**Comment by Maiqel** [comment]
This issue was fixed in commit cc6c69251d4150c5c16b9fc2b381cd821f451ba7 .
I just have it tweaked, and works for Ubuntu 20.04, kernel 5.8.0-59.
